src/og/entity/GeoObjectHandler.js

```diff
@@ -0,0 +1,1106 @@
+            }
+            if (!this._renderer.handler.programs.geo_object_picking) {
+                this._renderer.handler.addProgram(shaders.geo_object_picking());
+            }
+        }
+    }
+
+    setRenderNode(renderNode) {
+        this._renderer = renderNode.renderer;
+        this._planet = renderNode;
+        this.initProgram();
+    }
+
+    setRenderer(planet) {
+        super.setRenderer(planet);
+    }
+
+    _addInstancedGeoObjectToArray(geoObject) {
+        var itemSize = 3;
+
+        var tag = geoObject.tag,
+            alreadyAdded = this._instancedTags.has(tag);
+
+        if (!alreadyAdded) {
+            this._instancedTags.set(tag, {
+                vCounts: geoObject._verticesCount * itemSize,
+                iCounts: 1,
+                iSize: geoObject._indices.length,
+                index: this._instancedTags.size
+            });
+        } else {
+            const prevState = this._instancedTags.get(tag),
+                nextCount = prevState.iCounts + 1;
+            this._instancedTags.set(tag, {
+                vCounts: geoObject._verticesCount * itemSize,
+                iCounts: nextCount,
+                iSize: prevState.iSize,
+                index: prevState.index
+            });
+        }
+        const tagData = this._instancedTags.get(tag),
+            ti = tagData.index;
+
+        /**
+         * mark object by index for recalc indices array
+         */
+        geoObject._tagIndex = tagData.iCounts - 1;
+        if (geoObject._visibility) {
+            if (!alreadyAdded) {
+                this._vertexArr[ti] = concatArrays(
+                    this._vertexArr[ti],
+                    setParametersToArray(
+                        [],
+                        0,
+                        geoObject._verticesCount * itemSize,
+                        geoObject._verticesCount * itemSize,
+                        ...geoObject._vertices
+                    )
+                );
+                this._normalsArr[ti] = concatArrays(
+                    this._normalsArr[ti],
+                    setParametersToArray(
+                        [],
+                        0,
+                        geoObject._verticesCount * itemSize,
+                        geoObject._verticesCount * itemSize,
+                        ...geoObject._normals
+                        // ...getTriangleNormals(geoObject._vertices, geoObject._indices)
+                    )
+                );
+            }
+        } else {
+            this._vertexArr[ti] = concatArrays(
+                this._vertexArr[ti],
+                setParametersToArray([], 0, geoObject._verticesCount * itemSize, 1, 0)
+            );
+        }
+        var x = geoObject._positionHigh.x,
+            y = geoObject._positionHigh.y,
+            z = geoObject._positionHigh.z,
+            w;
+
+        this._positionHighArr[ti] = concatArrays(
+            this._positionHighArr[ti],
```

Comment:
I believe there is a bug, @pavletto check first array element, it's empty. I suppose it concerns the all arrays.

Reply:
`concatArray` has empty array in first parameter by default

Reply:
Oh, I think I got it. I'm reserved first element of arrays for uninstansed objects. Maybe it's unnecessary now.

Reply:
Fixed